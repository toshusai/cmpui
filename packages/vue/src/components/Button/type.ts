export const ButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type ButtonVariant = (typeof ButtonVariant)[keyof typeof ButtonVariant];

export const ButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type ButtonSize = (typeof ButtonSize)[keyof typeof ButtonSize];

export type ButtonProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
};
