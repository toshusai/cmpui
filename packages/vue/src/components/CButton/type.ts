export const CButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type CButtonVariant =
  (typeof CButtonVariant)[keyof typeof CButtonVariant];

export const CButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type CButtonSize = (typeof CButtonSize)[keyof typeof CButtonSize];

export type CButtonProps = {
  variant?: CButtonVariant;
  size?: CButtonSize;
  disabled?: boolean;
};
