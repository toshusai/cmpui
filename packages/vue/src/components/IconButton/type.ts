export const IconButtonSize = {
  S: "S",
  M: "M",
} as const;

export type IconButtonSize =
  (typeof IconButtonSize)[keyof typeof IconButtonSize];

export type IconButtonProps = {
  bordered?: boolean;
  disabled?: boolean;
  rounded?: boolean;
  selected?: boolean;
  size?: IconButtonSize;
};
