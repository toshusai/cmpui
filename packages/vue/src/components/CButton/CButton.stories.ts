import type { Meta, StoryObj } from "@storybook/vue3";
import CButton from "./CButton.vue";
import { h } from "vue";

const meta = {
  title: "inputs/CButton",
  tags: ["autodocs"],
  component: CButton,
} satisfies Meta<typeof CButton>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    default: "Button",
    variant: "primary",
  },
};

export const Variants: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary" }, () => "Primary"),
      h(CButton, { variant: "secondary" }, () => "Secondary"),
      h(CButton, { variant: "transparent" }, () => "Transparent"),
    ]),
};

export const Sizes: Story = {
  render: () =>
    h(() => [
      h(CButton, { size: "M" }, () => "Medium"),
      h(CButton, { size: "S" }, () => "Small"),
    ]),
};

export const Disabled: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary", disabled: true }, () => "Primary"),
      h(CButton, { variant: "secondary", disabled: true }, () => "Secondary"),
    ]),
};

export const Bordered: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary", bordered: true }, () => "Primary"),
      h(CButton, { variant: "secondary", bordered: true }, () => "Secondary"),
    ]),
};
