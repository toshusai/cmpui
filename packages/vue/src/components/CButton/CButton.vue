<script setup lang="ts">
import { CButtonProps } from "@toshusai/cmpui-core";
import { VNode } from "vue";

withDefaults(
  defineProps<
    CButtonProps & {
      is?: string | VNode;
    }
  >(),
  {
    is: "button",
    bordered: undefined,
    disabled: undefined,
    size: undefined,
    variant: undefined,
  },
);
</script>

<template>
  <component
    :is="is"
    class="cmpui_button"
    :data-bordered="bordered"
    :data-size="size"
    :data-variant="variant"
    :disabled="disabled"
    v-bind="$attrs"
  >
    <slot></slot>
  </component>
</template>
