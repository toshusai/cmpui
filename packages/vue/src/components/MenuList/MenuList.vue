<script setup lang="ts">
import { onMounted, onUnmounted, ref } from "vue";
import { useKeyboardNavigation } from "../../lib/menuItem";

const props = withDefaults(
  defineProps<{
    defaultValue?: string;
  }>(),
  {
    defaultValue: undefined,
  },
);

const emit = defineEmits<{
  (_e: "select", v: string): void;
}>();

const el = ref<HTMLDivElement | null>(null);

let cleanup = () => {};

onMounted(() => {
  if (!el.value) return;
  setTimeout(() => {
    el.value?.focus();
  }, 0);
  cleanup = useKeyboardNavigation(
    el.value,
    (value) => {
      emit("select", value);
    },
    {
      defaultValue: props.defaultValue,
    },
  );
});

onUnmounted(() => {
  cleanup();
});
</script>

<template>
  <div ref="el" class="cmpui_menu-list" role="listbox" tabindex="-1">
    <slot></slot>
  </div>
</template>
