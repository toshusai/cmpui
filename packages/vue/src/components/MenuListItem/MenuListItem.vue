<script setup lang="ts">
import ListItem from "../ListItem/ListItem.vue";

withDefaults(
  defineProps<{
    rounded?: boolean;
    selected?: boolean;
    value?: string;
    size?: "S" | "M";
  }>(),
  {
    size: "M",
    value: undefined,
  },
);
</script>

<template>
  <ListItem
    :data-size="size"
    :data-rounded="rounded"
    :aria-selected="selected"
    :data-value="value"
    tabindex="0"
    v-bind="$attrs"
  >
    <slot></slot>
  </ListItem>
</template>
