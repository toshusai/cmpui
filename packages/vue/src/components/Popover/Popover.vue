<script setup lang="ts">
import { CSSProperties, ref, watch } from "vue";
import { setPopover } from "../../lib/popover";

const divRef = ref<HTMLElement | null>(null);

const props = defineProps<{
  trigger: HTMLElement | null;
  show: boolean;
  disabledTriggerClickClose?: boolean;
}>();

const emit = defineEmits<{
  (_e: "close"): void;
}>();

const style = ref<CSSProperties>({
  display: "none",
  position: "fixed",
  top: "0",
  left: "0",
});

watch(
  () => [props.trigger, divRef.value, props.show],
  () => {
    if (!props.trigger) return;
    if (!divRef.value) return;
    const { cleanUp, open } = setPopover(
      divRef.value,
      props.trigger,
      () => {
        emit("close");
      },
      (pos) => {
        if (pos.left !== undefined && pos.top !== undefined) {
          style.value.transform = `translate(${pos.left}px, ${pos.top}px)`;
        }
        if (pos.maxWidth !== undefined) {
          style.value.maxWidth = `${pos.maxWidth}px`;
        }
        if (pos.maxHeight !== undefined) {
          style.value.maxHeight = `${pos.maxHeight}px`;
        }
        style.value.display = "";
      },
      {
        disabledTriggerClickClose: props.disabledTriggerClickClose,
      },
    );

    const handleOpenChange = (show: boolean) => {
      if (show) {
        open();
      } else {
        cleanUp();
        style.value.display = "none";
      }
    };

    handleOpenChange(props.show);

    watch(() => props.show, handleOpenChange);
  },
);
</script>

<template>
  <div
    ref="divRef"
    class="cmpui_float-box__root"
    tabindex="-1"
    v-bind="$attrs"
    :style="style"
  >
    <slot></slot>
  </div>
</template>
