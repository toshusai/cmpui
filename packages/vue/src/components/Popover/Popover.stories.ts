import type { Meta, StoryObj } from "@storybook/vue3";
import { ref } from "vue";
import CButton from "../CButton/CButton.vue";
import Popover from "./Popover.vue";
import MenuList from "../MenuList/MenuList.vue";
import MenuListItem from "../MenuListItem/MenuListItem.vue";

const meta = {
  component: Popover,
} satisfies Meta<typeof Popover>;

export default meta;
type Story = StoryObj<typeof meta>;

const html = String.raw;

export const Default: Story = {
  render: () => ({
    setup() {
      const trigger = ref<InstanceType<typeof CButton> | null>(null);

      const show = ref(false);

      const selected = ref<string | null>(null);

      const handleSelect = (v: string) => {
        selected.value = v;
        show.value = false;
      };

      return {
        trigger,
        show,
        selected,
        handleSelect,
      };
    },
    components: { Popover, CButton, MenuList, MenuListItem },
    template: html`
      <CButton ref="trigger" @click="show = true">
        {{ selected || "Select" }}
      </CButton>
      <Popover
        v-if="show"
        :trigger="trigger?.$el"
        :show="show"
        @close="show = false"
      >
        <MenuList @select="handleSelect">
          <MenuListItem is="button" value="Alpha">Alpha</MenuListItem>
          <MenuListItem is="button" value="Beta">Beta</MenuListItem>
          <MenuListItem is="button" value="Gamma">Gamma</MenuListItem>
          <input type="text" />
        </MenuList>
      </Popover>
    `,
  }),
};

export const Scroll: Story = {
  render: () => ({
    setup() {
      const trigger = ref<InstanceType<typeof CButton> | null>(null);

      const show = ref(false);

      const selected = ref<string | null>(null);

      const handleSelect = (v: string) => {
        selected.value = v;
        show.value = false;
      };

      return {
        trigger,
        show,
        selected,
        handleSelect,
      };
    },
    components: { Popover, CButton, MenuList, MenuListItem },
    template: html`
      <div style="height: 400px; overflow: auto;"></div>
      <CButton ref="trigger" @click="show = true">
        {{ selected || "Select" }}
      </CButton>
      <Popover
        v-if="show"
        :trigger="trigger?.$el"
        :show="show"
        @close="show = false"
      >
        <MenuList @select="handleSelect">
          <MenuListItem is="button" value="Alpha">Alpha</MenuListItem>
          <MenuListItem is="button" value="Beta">Beta</MenuListItem>
          <MenuListItem is="button" value="Gamma">Gamma</MenuListItem>
          <input type="text" />
        </MenuList>
      </Popover>
      <div style="height: 400px; overflow: auto;"></div>
    `,
  }),
};
