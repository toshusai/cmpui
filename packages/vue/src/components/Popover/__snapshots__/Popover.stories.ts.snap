// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Popover Default smoke-test 1`] = `
<button class="cmpui_button"
        data-variant="primary"
        data-size="M"
>
  Select
</button>
`;

exports[`components/Popover Scroll smoke-test 1`] = `
<div style="height: 400px; overflow: auto;">
</div>
<button class="cmpui_button"
        data-variant="primary"
        data-size="M"
>
  Select
</button>
<div style="height: 400px; overflow: auto;">
</div>
`;
